harness/determined/core/_context.py

```diff
@@ -154,9 +172,11 @@ def init(
         preempt_mode (``core.PreemptMode``, optional): Configure the calling pattern for the
             ``core_context.preempt.should_preempt()`` method.  See
             :class:`~determined.core.PreemptMode` for more detail.  Defaults to ``WorkersAskChief``.
-        checkpoint_storage (``str``, optional): A directory path or a cloud storage URI of the form
-            ``s3://<bucket>[/<prefix>]`` (AWS) or ``gs://<bucket>[/<prefix>]`` (GCP). This should
-            only be used when IAM permissions can be assumed.
+        checkpoint_storage (``Union[str, dict]``, optional): A directory path or a cloud storage URI
+            of the form ``s3://<bucket>[/<prefix>]`` (AWS) or ``gs://<bucket>[/<prefix>]`` (GCP).
+            This should only be used when IAM permissions can be assumed. Except in the case of a
+            shared file system, a dictionary may also be given with fields documented in the
+            Experiment Configuration Reference.
```

Comment:
For technical writing, I think frontloading your exception, as in "Except for ..., you may also ...", should be avoided.

It means the same but is less clear and direct than saying "You may also ..., except for ...".

Also, I would avoid saying "shared file system" here.  I'd write the exception as: "except in the case of type:shared_fs" or something.  Reason: "shared file system" is a pretty general idea, when we're really talking about a very specific idea: checkpoint storage configs with `type: shared_fs`.  You might have bind-mounted a shared filesystem somewhere else into your container, but if that's not what you configured your checkpoint storage on, it isn't relevant.

Last thing: "given with fields documented in the Experiment Configuration Reference" is too broad.  Should the pass the whole experiment config in?  Or just the `checkpoint_storage` config?

My proposal:

> A directory path or a ... can be assumed.  You may also pass a dictionary matching the checkpoint_storage field of the experiment config, with the exception that ``type: shared_fs`` configs are not allowed.